pkg/webhook/resourceusage/util.go

```diff
@@ -0,0 +1,245 @@
+func parseJavaMemoryString(str string) (int64, error) {
+	lower := strings.ToLower(str)
+	if matches := javaStringPattern.FindStringSubmatch(lower); matches != nil {
+		value, err := strconv.ParseInt(matches[1], 10, 64)
+		if err != nil {
+			return 0, err
+		}
+		suffix := matches[2]
+		if multiplier, present := javaStringSuffixes[suffix]; present {
+			return multiplier * value, nil
+		}
+	} else if matches = javaFractionStringPattern.FindStringSubmatch(lower); matches != nil {
+		value, err := strconv.ParseFloat(matches[1], 64)
+		if err != nil {
+			return 0, err
+		}
+		suffix := matches[2]
+		if multiplier, present := javaStringSuffixes[suffix]; present {
+			return int64(float64(multiplier) * value), nil
+		}
+	}
+	return 0, fmt.Errorf("could not parse string '%s' as a Java-style memory value. Examples: 100kb, 1.5mb, 1g", str)
+}
+
+// Logic copied from https://github.com/apache/spark/blob/c4bbfd177b4e7cb46f47b39df9fd71d2d9a12c6d/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala
+func memoryRequiredForSparkPod(spec so.SparkPodSpec, memoryOverheadFactor *string, appType so.SparkApplicationType, replicas int64) (int64, error) {
+	var memoryBytes int64
+	if spec.Memory != nil {
+		memory, err := parseJavaMemoryString(*spec.Memory)
+		if err != nil {
+			return 0, err
+		}
+		memoryBytes = memory
+	} else {
+		memoryBytes = defaultMemoryBytes
+	}
+	var memoryOverheadBytes int64
+	if spec.MemoryOverhead != nil {
+		overhead, err := parseJavaMemoryString(*spec.MemoryOverhead)
+		if err != nil {
+			return 0, err
+		}
+		memoryOverheadBytes = overhead
+	} else {
+		var overheadFactor float64
+		if memoryOverheadFactor != nil {
+			overheadFactorScope, err := strconv.ParseFloat(*memoryOverheadFactor, 64)
+			if err != nil {
+				return 0, err
+			}
+			overheadFactor = overheadFactorScope
+		} else {
+			if appType == so.JavaApplicationType {
+				overheadFactor = defaultMemoryOverhead
+			} else {
+				overheadFactor = nonJvmDefaultMemoryOverhead
+			}
+		}
+		memoryOverheadBytes = int64(math.Max(overheadFactor*float64(memoryBytes), minMemoryOverhead))
+	}
+	return (memoryBytes + memoryOverheadBytes) * replicas, nil
+}
+
+func resourceUsage(spec so.SparkApplicationSpec) (ResourceList, error) {
+	driverMemoryOverheadFactor := spec.MemoryOverheadFactor
+	executorMemoryOverheadFactor := spec.MemoryOverheadFactor
+	if spec.Driver.MemoryOverhead != nil {
+		driverMemoryOverheadFactor = spec.Driver.SparkPodSpec.MemoryOverhead
+	}
+	if spec.Executor.MemoryOverhead != nil {
+		executorMemoryOverheadFactor = spec.Executor.SparkPodSpec.MemoryOverhead
+	}
+
+	driverMemory, err := memoryRequiredForSparkPod(spec.Driver.SparkPodSpec, driverMemoryOverheadFactor, spec.Type, 1)
+	if err != nil {
+		return ResourceList{}, err
+	}
+
+	var instances int64
+	if spec.Executor.Instances != nil {
+		instances = int64(*spec.Executor.Instances)
+	} else {
+		instances = 1
```

Comment:
You can assign 1 to `instances` when initializing it so you don't need the `else`.